File: Shortest Path/Dijkstra/1162번 - 도로포장/1162.py
import sys
import heapq

input = sys.stdin.readline

INF = 10**15

n, m, k = map(int, input().split())
edges = list([] for _ in range(n + 1))
for _ in range(m):
    a, b, c = map(int, input().split())
    edges[a].append((b, c))
    edges[b].append((a, c))

def solve():
    dp = [[INF] * (n + 1) for _ in range(k + 1)]
    hq = []
    heapq.heappush(hq, [0, 1, 0])
    
    while hq:
        weight, node, cnt = heapq.heappop(hq)
        
        if dp[cnt][node] < weight:
            continue
        
        for nxt_node, nxt_weight in edges[node]:
            nxt_weight += weight
            
            if dp[cnt][nxt_node] > nxt_weight:
                dp[cnt][nxt_node] = nxt_weight
                heapq.heappush(hq, [nxt_weight, nxt_node, cnt])
            
            if cnt + 1 <= k and dp[cnt + 1][nxt_node] > weight:
                dp[cnt + 1][nxt_node] = weight
                heapq.heappush(hq, [weight, nxt_node, cnt + 1])
                
    print(min(dp[i][n] for i in range(1, k + 1)))

if __name__ == '__main__':
    solve()

File: Shortest Path/Dijkstra/1162번 - 도로포장/1162.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

#define INF 1e15
#define ll long long
#define X first
#define Y second

int n, m, k;
vector<pair<int, int>> edges[10001];
ll dp[21][10001];

void input() {
    cin >> n >> m >> k;
    for (int i = 0, a, b, c; i < m; i++) {
        cin >> a >> b >> c;
        edges[a].push_back({b, c});
        edges[b].push_back({a, c});
    }
}

ll dijkstra() {
    fill_n(&dp[0][0], 21 * 10001, INF);
    priority_queue<pair<pair<ll, int>, int> > pq;
    pq.push({{0, 1}, 0});
    dp[0][1] = 0;

    while (!pq.empty()) {
        ll cost = -pq.top().X.X;
        int node = pq.top().X.Y;
        int cnt = pq.top().Y;
        pq.pop();
        
        if (dp[cnt][node] < cost) continue;

        for (auto &edge : edges[node]) {
            int nxt_node = edge.X;
            ll nxt_cost = cost + edge.Y;

            if (dp[cnt][nxt_node] > nxt_cost) {
                dp[cnt][nxt_node] = nxt_cost;
                pq.push({{-nxt_cost, nxt_node}, cnt});
            }

            if (dp[cnt + 1][nxt_node] > cost && cnt + 1 <= k) {
                dp[cnt + 1][nxt_node] = cost;
                pq.push({{-cost, nxt_node}, cnt + 1});
            }
        }
    }

    ll ans = INF;
    for (int i = 0; i <= k; i++) ans = min(ans, dp[i][n]);
    return ans;
}

void solve() {
    cout << dijkstra() << '\n';
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    input();
    solve();

    return 0;
}
